modules/battery_manager/ajax/add_or_edit_entry.php

```diff
@@ -0,0 +1,207 @@
+
+    // Make sure new entry is active
+    $form_data['Active'] = 'Y';
+
+    // Check for duplicates on the back-end
+    if (checkForDuplicate() !== null) {
+        header('HTTP/1.1 400 Bad Request');
+        exit("There exists a duplicate entry in the Test Battery.");
+    } else {
+        // Add entry to Test Battery
+        try {
+            $db->insert('test_battery', $form_data);
+        } catch (DatabaseException $e) {
+            showError("Could not add entry to the Test Battery. Please try again!");
+        }
+    }
+}
+
+/**
+ * Handle updates in the test battery
+ *
+ * @throws DatabaseException
+ *
+ * @return void
+ */
+function editEntry()
+{
+    $db   =& \Database::singleton();
+    $user =& User::singleton();
+    if (!$user->hasPermission('battery_manager_edit')) {
+        header("HTTP/1.1 403 Forbidden");
+        exit;
+    }
+
+    // Get ID of edited entry
+    $entryID = $_POST['id'];
+
+    // Retrieve values entered by user
+    $form_data = getFormData();
+
+    // Get active status of edited entry
+    $form_data['Active'] = $_POST['active'];
+
+    // Check for duplicates on the back-end
+    if (checkForDuplicate() !== null) {
+        header('HTTP/1.1 400 Bad Request');
+        exit("There exists a duplicate entry in the Test Battery.");
+    } else {
+        // Update entry in Test Battery
+        try {
+            $db->update('test_battery', $form_data, ['ID' => $entryID]);
+        } catch (DatabaseException $e) {
+            showError(
+                "Could not update entry "+$entryID+" in the Test Battery."
+                ." Please try again!"
+            );
+        }
+    }
+}
+
+/**
+ * Retrieve values entered by user
+ *
+ * @return array
+ */
+function getFormData()
+{
+    $form_data = array(
+                  'Test_name'    => $_POST['instrument'] ?? null,
+                  'AgeMinDays'   => $_POST['ageMinDays'] ?? null,
+                  'AgeMaxDays'   => $_POST['ageMaxDays'] ?? null,
+                  'Stage'        => $_POST['stage'] ?? null,
+                  'SubprojectID' => $_POST['subproject'] ?? null,
+                  'Visit_label'  => $_POST['visitLabel'] ?? null,
+                  'CenterID'     => $_POST['forSite'] ?? null,
+                  'firstVisit'   => $_POST['firstVisit'] ?? null,
+                  'instr_order'  => $_POST['instrumentOrder'] ?? null,
+                 );
+
+    // Convert null strings to nulls
+    foreach ($form_data as $key => $value) {
+        if ($value === "null") {
+            $form_data[$key] = null;
+        }
+    }
+
+    return $form_data;
+}
+
+/**
+ * Utility function to return errors from the server
+ *
+ * @param string $message error message to display
+ *
+ * @return void
+ */
+function showError($message)
+{
+    if (!isset($message)) {
+        $message = 'An unknown error occurred!';
+    }
+    header('HTTP/1.1 500 Internal Server Error');
+    header('Content-Type: application/json; charset=UTF-8');
```

Comment:
This doesn't seem to be return json? (or doing anything with message?)

Reply:
if i add exit($message) in the last line, the error message is passed to the front-end. however, i was told to avoid die/exit. is there a better to return the error?